### ISSUE TITLE
UI/UX for Application Sharing

### ISSUE THREAD

**Issue by chrsamp** [bcgov/api-services-portal#878]

**Comment by odyeung** [comment]
@ikethecoder bump! questions from nathalia...

**Comment by nathaliamorales** [comment]
im meeting with Devs today to discuss whats technically possible from the backend before heading with the mockups.

**Comment by odyeung** [comment]
@ikethecoder bump again -- @nirajCITZ nathalia is going to reach out to you just in case you can help unblock her

**Comment by nathaliamorales** [comment]
I spoke @nirajCITZ and he will run some tests to give advice on what would be the best way to share app ownership. If we can support sharing with more than 1 user at the time, max 2, or once at a time.

**Comment by nirajCITZ** [comment]
Checked keycloak and postgrace db to check how application is maped with the user but it requires some more investigation to verify whether it's feasible to share application with another user

**Comment by odyeung** [comment]
@ikethecoder @Jonesy need you to see what is being called!

**Comment by nathaliamorales** [comment]
@ikethecoder yes chris mentioned that we wont have different roles. Everyone would be "owner" and all have the same rights.

Im not sure what the concept "team" implies and how it would simplify the administration or if its needed.

**Comment by nathaliamorales** [comment]
im reviewing with Aidan some ideas he has around the way we share permission in and org level. We are taking a step back and Aidan is helping me understand the current APS structure and an example of how users interact with the portal on different stages. the outcome will be something similar to a "mental map"

**Comment by nathaliamorales** [comment]
Met with Aidan to understand the API creation process, now im editing the "map". Will be shared when finished. We are having continuation for the App. sharing conversation on Thursday

**Comment by nathaliamorales** [comment]
I met with Aidan today to get feedback on the 'application sharing' designs. As a result of this meeting, I will be doing some modifications to the designs considering:

- Ownership will be shared as an invite (an email with a one time use link) that will have certain time (for now it was discussed 2 days) for the recipient to login the APS portal in order to accept it.
- We cant make sure the recipient received the email, we can only confirm it was sent
- The only validation we can make it's if the value has or not an email format (standard email validation)
- We wont show "[application name] updated" success message as we do when the user changes the application name or description. This is because when the invite its sent it doesn't mean they now are owners. they need to accept by logging in. The success message will be more specific (something more like "invites sent"). For now the generic success message for other changes will remain as is.
- 'Invites' will have 3 states: waiting, accepted, expired. Thinking about adding a re send button.
- Nathalia to consider separating the invite sharing from the 'edit application modal'. Probably it will make the interaction easier. Needs to define if sharing ownership is considering an edition to the application, or if its separate. If its separate, then a 3rd option will be added to the "more action" button (the 2 dots)

**Comment by nathaliamorales** [comment]
I met with Aidan, Virgina, James and Josh for the last round of feedback. These were the agreements:

- Emails need to be provided in order to send the invitation links. For expired or waiting for response status, we will show the email of the person. When users accept the invitation, we will show the email if we have it. if not we show only the display name.
- We will allow deleting an item with the first clic on the delete button. we will add a undo banner similar to gmail or outlook to allow users recover from error without adding more clics to show a modal confirmation
- i will create the flow for the recipient who wants to accept the invite

This will continued on sprint 85

**Comment by nathaliamorales** [comment]
@ikethecoder  how the recipient's flow looks like? 
I imagine something like: the person receives an email that has a login button/link which take them to the portal. Right after they login i guess there would be a message saying you are now co-owner of application 123. If the user goes to my applications they will see the application there. 
I guess they will see exactly the same as the user who invited them, when clicking on "manage application". They will see the list of owners, the option to resend, remove and add more owners.
Can you confirm if this is right? or how this flow would be? I hope I can map that flow this week before I leave :o

**Comment by ikethecoder** [comment]
Yes, that is correct.  The comment "This will also require an effort to keep track of the invite activity in the gateway activity page. E.g. "Jane Lucas invited [jhon@gmail.com](mailto:jhon@gmail.com) to share ownership"" - this is not quite true though as the My Applications does not have anything to do with a Namespace/Gateway, so there is no Activity.  I think we had said that we would display the invitation email on the Manage Ownership dialog so that they are tied together (?)

**Comment by nathaliamorales** [comment]
1. @ikethecoder when you say "that is correct" you refer to the recipient's flow i mentioned in the comment below right?
2. About the track in the activity. I dont fully remember but i think we mentioned something about having the track in the activity, but if the application is not related to a gateway then it shouldn't be there. So i will remove this point form the developer notes i left in the figma file.
3. Yes we said we will keep always the email used for the invite to avoid confusion. First it was said we would take whatever info the invited person used to login (which could be a username or a different email) an accept the invitation. But then i recommended to always keep visible (in the manage ownership modal) the same email used for the invite, EVEN if the person logged in with different email or with username.  Are we on the same page here?

**Comment by ikethecoder** [comment]
Re point 1 - yes
Re point 3 - will the manage ownership show both the original email and the accepted user's info?  Thinking that over time, updated user info will come from the identity provider (BCSC, IDIR, Github, etc) so that should be visible somewhere.. the original email will be irrelevant over time.

**Comment by nathaliamorales** [comment]
After our collaborative discussion earlier today, we have decided to revert to the original plan mentioned on November 16th: _"For expired or waiting for response status, we will show the email of the person. When users accept the invitation, we will show the information we can get when the user clicks on the link and log in, which could be a different email from the invite or a display name. This is because users can receive the invite to a specific email but then accept the invite and loging in with different credentials. If we can get the email the user used to login/accept the invite we will show it. If the user doesnt login with an email then we will show the display name."_


During today's discussion, we explored two possible solutions to prevent users from getting confused in case the email to which the invite was sent differs from the information when the invite is accepted:

Add an option somewhere on the 'edit owner' list element to view the original email to which the invite was sent and when.
Introduce an activity page for each application (similar to the one in namespaces) where users can review all activities related to the application sharing, such as who was invited, when, by whom, deletions, etc. While this could be considered a sophisticated feature, it may be highly desired by users, as Aidan mentioned.
Implementing an additional info button for mouse-over to display the email and date when the invite was sent may clutter the element, considering we already have an icon (to indicate the state), a resend button, and a delete button. To maintain a clean interface, my recommendation, for now, is getting the provider identity, and once implemented and in use we can gather feedback from real users to assess:

- Whether it is confusing to see the invite identity changing or not.
- Whether an activity page would be a desired feature.

**Comment by nathaliamorales** [comment]
Designs are finished. Find the screens and the development notes in the figma file, in the 'ready for development page'. This ticket can be closed or assigned to a developer for implementation :)

**Comment by rustyjux** [comment]
is this thing on?

**Comment by odyeung** [comment]
Yes sir!

**Comment by veep-22** [comment]
RocketChat request from Willem Mulder on May 15 2024 - emerging requirement for SAWSx and DriveBC
Aidan Cope @aidan.cope
Owner
11:28 AM
Hi Willem Mulder, at the moment the only option is for the person replacing you to go through creating an application and requesting new credentials again. But we have completed the UX work on this change ("Application Sharing") and the development work is in our backlog (see https://github.com/bcgov/api-services-portal/issues/127). cc'ing our product owner Virginia Vella for prioritization review.

Willem Mulder
[10:54 AM](https://chat.developer.gov.bc.ca/channel/aps-ops?msg=CuysydeTsqZzo66ST)
I am working on the new DriveBC site and as part of that we are pulling in weather data from SAWSx through the gateway. I used my IDIR to login to the API Services portal and created an application + requested access with my account.
My question is, what happens if I were to move to another role and would no longer be managing that API connection? Is there a way for me to share my application with others? Should we be using a service account to set up this connection? What is the best practice?
